### ISSUE TITLE
Feature request: no "revert" on ESC in text input

### ISSUE THREAD

**Issue by aardappel** [ocornut/imgui#6578]
**Version/Branch of Dear ImGui:**

Version: 4/11/2023
Branch: docking

**Back-end/Renderer/Compiler/OS**

Back-ends: OpenGL
Compiler: VS2023
Operating System: Windows

**My Issue/Question:**

Pressing ESC in a text input reverts edits, and there seems to be no flag to control this behavior (only to make it clear, which is worse). I'd like ESC to only deselect.

Scenario: ESC in game toggles between game and menus (created using Dear ImGui). User makes a change in a string widget, presses ESC to go back to the game, and is surprised their edits were lost.

We now have to explain them to click outside the text widget before going back to the game to register their changes. That should not be necessary.

Toggling menus/UI with ESC is very common UI in games, so it also meaning "revert" is problematic :)

**Comment by nicolasnoble** [comment]
Can't you just not send the keypress to imgui?

**Comment by aardappel** [comment]
> Can't you just not send the keypress to imgui?

Hah I feel silly now, I forgot I even have control over that. Bit of a hack, since normally events go straight to imgui, and I only process what it doesn't want, but I suppose I can pre-filter. I guess I'll do that.. false alarm, then!

**Comment by ocornut** [comment]
> Can't you just not send the keypress to imgui?

I would not recommend this as it would requires a close link between your backend code and your application code.

A few versions I added API (still in imgui_internal.h) that should solve this,

This Demo code in this branch:
https://github.com/ocornut/imgui/compare/master...features/demo_input_owner_and_routing
(this is a branch because it adds an `#include "imgui_internal.h"` in the demo but those API will be moved to public realm in some time)

There are various way to achieve this but I think you can call `ImGui::SetKeyOwner(ImGuiKey_Escape, ImGui::GetID("")); // any id` 
or `ImGui::SetKeyOwner(ImGuiKey_Escape, 0, ImGuiInputFlags_LockThisFrame)`

**Comment by ocornut** [comment]
I have complemented the demo with this exact scenario:
```cpp
if (ImGui::TreeNode("6. Disable Escape key from an InputText()"))
{
    static char buf[9];
    ImGui::InputText("Text", buf, IM_ARRAYSIZE(buf));

    // - If you don't need to use the key, you can use 'owner_id=0', 'flags=ImGuiInputFlags_LockXXX'
    //   as a convenience to hide the key from everyone.
    // - If you need to use the key yourself, you need to use any arbitrary ID, and then use this ID to read the key.
    //   e.g. ImGui::SetKeyOwner(ImGuiKey_Escape, ImGui::GetID("robber")); + later use same ID to access the key.
    if (ImGui::IsItemActive())
        ImGui::SetKeyOwner(ImGuiKey_Escape, 0, ImGuiInputFlags_LockUntilRelease);
    ImGui::TreePop();
}
```

**Comment by aardappel** [comment]
@ocornut thanks!

Maybe this issue should be left open then until those functions arrive in the main branches?

**Comment by ocornut** [comment]
Not necessary, they will be advertised and demo-ed and all threads linked accordingly.

**Comment by ocornut** [comment]
>  arrive in the main branches?

Sorry for the confusion: this is ALREADY in the main branch. Only the DEMO code isn't because the API is currently in imgui_internal.h. But that specific API call will not change.

**Comment by ocornut** [comment]
Have you read the comment/section?

```
// - If you don't need to use the key, you can use 'owner_id=0', 'flags=ImGuiInputFlags_LockXXX'
//   as a convenience to hide the key from everyone.
// - If you need to use the key yourself, you need to use any arbitrary ID, and then use this ID to read the key.
//   e.g. ImGui::SetKeyOwner(ImGuiKey_Escape, ImGui::GetID("robber")); + later use same ID to access the key.
if (ImGui::IsItemActive())
    ImGui::SetKeyOwner(ImGuiKey_Escape, 0, ImGuiInputFlags_LockUntilRelease);
```

**Comment by ocornut** [comment]
If you use `SetKeyOwner(ImGuiKey_Escape, ImGui::GetID("foobar"))` then:

- `IsKeyPressed(ImGuiKey_Escape)` will return true. That's the "key owner unaware api".
- `IsKeyPressed(ImGuiKey_Escape, ImGui::GetID("foobar"))` will return true. That's the "key owner aware api".
- `IsKeyPressed(ImGuiKey_Escape, ImGui::GetID("foobar2"))` will return false. That's the "key owner aware api".

All standard widgets are key-owner aware.

**Comment by ocornut** [comment]
That said, regardless of the more generic key owner system (which can be applied to this but also many other things), I think it may make sense to add an explicit `InputText()` flag e.g. `ImGuiInputTextFlags_EscapeNoRevert` (name to be determined) to simply disable that one behavior at it may be commonly desirable.

But you should be able to get it to work now with aforementioned calls.

**Comment by ocornut** [comment]
I'll add a little more details (as those API are currently not public, were mentioned in variety of posts to solve specific problems but are still lacking a big post):

The idea here is that:
- An active `InputText()` tries to claim ownership of the Escape key.
- Here my suggested solution above is to steal that ownership after the `InputText()` runs, but only if it was active.

You could however take the problem the opposite way: 
- Your game toggle could claim ownership of the Escape key (by calling `SetKeyOwner()` with its arbitrary ID).
- InputText() won't access it because it is input-owner aware.

However that second alternative would be less flexible as you would be stealing ESC from everyone else, not just that `InputText()`. So first solution seems preferable.

**Comment by aardappel** [comment]
I use SDL2 for all my keyboard handling, so I don't want to have to go to imgui just to poll ESC.

The structure I have: get event from SDL -> send to imgui -> process the SDL event unless the mouse/keyboard flag is false.

**Comment by ocornut** [comment]
Keyboard is normally captured by imgui when a widget is capture but you could call SetNextFrameWantCaptureKeyboard(false) when it is active, the problem is you would receive other keys as well. 

Perhaps you need to set another flag to coordinates that state with your app and only let ESC pass.

All of this discussion is assuming you still want dear imgui to handle some ESC events (eg closing popups or some other input texts). If you never ever want imgui to see ESC key then we are back to the first suggestion (“Can't you just not send the keypress to imgui”).

**Comment by aardappel** [comment]
Yeah for now it's just easier to not give the ESC to imgui at all.. if a popup was open and a user pressed ESC I bet they'd still expect to go back to the game, as that is what that key always does.

I personally wasn't aware ESC would undo edits/popups in any UI, and never used it that way.. will see if users end up missing it :)

Still, `ImGuiInputTextFlags_EscapeNoRevert` would be useful I think.

**Comment by FireFox2000000** [comment]
Having a flag would be useful for my use-case as I'm currently using the "escape" key as the "back" button shortcut for my custom-state management. What's happening at the moment is that if a user types in an input field and then hits the "escape" key to return to the previous state, the input field ends up reverting the data they just entered when it's expected that the data should be auto-saved as they exit from the menu. Though I can see how having that revert functionality would be useful on other input fields within different parts of my application.

**Comment by ocornut** [comment]
You can do as discussed in the thread and claim ownership of that key away from the InputTexr() widget, and process it yourself.

**Comment by FireFox2000000** [comment]
For further context, in my project I've got ImGUI running within a Unity application, which requires the Dear ImGUI native library, ImGUI.NET for C# bindings, and then lastly another wrapper on top of that to connect Unity to ImGUI.NET. This also means I don't have access to the api provided by imgui_internal, which `SetKeyOwner ` looks to be accessed from (I could be wrong, github search is being bit unreliable, at the very least it's not available in the current C# bindings I'm using). 
Although I could edit the native library for custom functionality, from a maintenance and upgradability perspective that path is far more unwieldy as it would require maintaining 3 separate custom forks, the native lib, C# bindings and then the Unity wrapper. 

Currently I'm just not sending the escape key to ImGUI at all for the time being, which appears to be working fine at the moment. So I don't actually have a problem at the moment, but I'm mainly just throwing my use-case in here in case it's deemed worth adding this flag down the line. 
Though imo it does seem a bit inconsistent how some features are publicly disableable (like undo redo through NoUndoRedo) and some are not like this one.
